File: questions/994.RottingOranges.go
package questions

func orangesRotting(grid [][]int) int {
	var queue [][]int
	time := 0
	fresh := 0

	rows := len(grid)
	cols := len(grid[0])

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] == 1 {
				fresh++
			}
			if grid[r][c] == 2 {
				queue = append(queue, []int{r, c})
			}
		}
	}

	directions := [][]int{
		{1, 0},
		{-1, 0},
		{0, 1},
		{0, -1},
	}

	for len(queue) > 0 && fresh > 0 {
		length := len(queue)
		for i := 0; i < length; i++ {
			popped := queue[0]
			queue = queue[1:]

			r, c := popped[0], popped[1]

			for _, direction := range directions {
				neighborRow := r + direction[0]
				neighborCol := c + direction[1]

				if neighborRow < 0 {
					continue
				}
				if neighborRow >= rows {
					continue
				}
				if neighborCol < 0 {
					continue
				}
				if neighborCol >= cols {
					continue
				}

				if grid[neighborRow][neighborCol] != 1 {
					continue
				}

				grid[neighborRow][neighborCol] = 2
				queue = append(queue, []int{neighborRow, neighborCol})
				fresh--
			}
		}
		time++
	}

	if fresh != 0 {
		return -1
	}

	return time
}

File: leetcode/994.RottingOranges.ts
function orangesRotting(grid: number[][]): number {
  const rows = grid.length
  const cols = grid[0].length
  const directions = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1],
  ]
  let time = 0
  let fresh = 0
  const queue: number[][] = []

  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (grid[r][c] === 1) fresh++
      if (grid[r][c] === 2) queue.push([r, c])
    }
  }

  while (queue.length && fresh > 0) {
    const length = queue.length
    for (let i = 0; i < length; i++) {
      const [row, col] = queue.shift()!
      for (const direction of directions) {
        const neighborRow = row + direction[0]
        const neighborCol = col + direction[1]

        if (grid[neighborRow]?.[neighborCol] !== 1) continue

        grid[neighborRow][neighborCol] = 2
        queue.push([neighborRow, neighborCol])
        fresh--
      }
    }
    time++
  }

  if (fresh !== 0) return -1

  return time
}

console.log(
  orangesRotting([
    [2, 1, 1],
    [1, 1, 0],
    [0, 1, 1],
  ]),
)
